### DISCUSSION TITLE
[Q] Access to http server pool events

### DISCUSSION THREAD

**Discussion by vasilevrv** [spiral/roadrunner-plugins#45] [Q&A] [answered]
Hello!

I need to add event listeners to http server pool for observe faults and send to Sentry. For example https://github.com/UPDG/roadrunner-sentry

How can i do it in v2?

**Comment by rustatian** [comment, marked as answer]
Hey @vasilevrv . This part of the RR2 left the same as was in the RR1 (but will be reworked in the future versions). You can provide your listener directly to the https://github.com/spiral/roadrunner-plugins/blob/master/server/plugin.go#L143 `NewWorkerPool` method.

**Reply by rustatian** [reply]
For example, this is a sample of how it works now (`http` plugin):
1. Listener: https://github.com/spiral/roadrunner-plugins/blob/master/http/plugin.go#L120
2. Attach a listener to the pool: https://github.com/spiral/roadrunner-plugins/blob/master/http/plugin.go#L153
The last arg in the `p.server.NewWorkerPool` method is variadic of options. One of the options is https://github.com/spiral/roadrunner/blob/master/pool/static_pool.go#L108 `AddListeners` option.

**Reply by rustatian** [reply]
This is an internal abstraction, not the public API (yet). Will be changed in near future with notice.

**Reply by vasilevrv** [reply]
> This is an internal abstraction, not the public API (yet). Will be changed in near future with notice.

Thanks for your reply!

I am using http plugin and I need to log worker crashes in Sentry. As far as I understand, it not possible now, rught?

https://github.com/spiral/roadrunner-plugins/blob/master/http/plugin.go#L146

**Reply by rustatian** [reply]
Yeah, this is not possible at the moment. We don't support sentry. But, if you create a feature_request and describe your proposal, we will take this into account for sure :)

**Reply by rustatian** [reply]
But, wait, do you use `http` plugin in your own bundle?

**Reply by vasilevrv** [reply]
@rustatian, Now I'm using a regular roadrunner server build (http and rpc with Symfony) in my project and I have a need to integrate Roadrunner with Sentry. I wanted to do it myself, downloaded the roadrunner-binary from github and added my plugin there. But, I ran into the fact that from the plugin I could not subscribe to the crash events of the worker. If you can tell me the way how i can implement it, I will be very grateful

**Reply by rustatian** [reply]
So, basically, you need to add your own listeners specified here: https://github.com/UPDG/roadrunner-sentry/blob/master/service.go#L25, to the http workers pool, am I right?

**Reply by vasilevrv** [reply]
@rustatian, yes!

**Reply by vasilevrv** [reply]
I understand (maybe I'm wrong) that this plugin is outdated and did not find a way to do it in the current RR2 version. This plugin does exactly what I need

**Reply by rustatian** [reply]
The easiest solution for you at the moment will be to fork the `http` plugin, add sentry DSN to its config and add your listeners to the pool here https://github.com/spiral/roadrunner-plugins/blob/master/http/plugin.go#L153 and here https://github.com/spiral/roadrunner-plugins/blob/master/http/plugin.go#L333.
Then replace our `http` plugin with your in the roadrunner-bundle.
In the near future, you'll be able to attach to the smt like central events bus for the particular event from any plugin.

**Reply by vasilevrv** [reply]
@rustatian, ok, thanks!

**Reply by vasilevrv** [reply]
> The easiest solution for you at the moment will be to fork the `http` plugin, add sentry DSN to its config and add your listeners to the pool here https://github.com/spiral/roadrunner-plugins/blob/master/http/plugin.go#L153 and here https://github.com/spiral/roadrunner-plugins/blob/master/http/plugin.go#L333.
> Then replace our `http` plugin with your in the roadrunner-bundle.
> In the near future, you'll be able to attach to the smt like central events bus for the particular event from any plugin.

is there an estimate date when it is planned to do?

**Reply by rustatian** [reply]
I guess 1-2 months. Here is the ticket: https://github.com/spiral/roadrunner/issues/496

**Reply by vasilevrv** [reply]
thanks!

**Reply by rustatian** [reply]
HTTP will not expose any events anymore (and will not expose in the future). You may create a simple HTTP middleware w/o binding to the HTTP plugin directly. You may have a look at the different middleware here: https://github.com/spiral/roadrunner-plugins/tree/master/http/middleware.
As basic advice on how to create a `sentry` middleware:
- Create a configuration -> example: https://github.com/spiral/roadrunner-plugins/blob/master/http/middleware/static/config.go
```yaml
http:
  sentry:
    dsn: ...
    user: ...
    other_options: ...
```
- Implement `Middleware` interface: https://github.com/spiral/roadrunner-plugins/blob/master/http/middleware/static/plugin.go#L99
- Create Init (mandatory), `Name` functions, and optional Serve/Stop.
- In the `Middleware` function, wrap the `http.ResponseWriter` interface, get all needed http codes, messages and build your logic based on this info. Wrapper sample: https://github.com/spiral/roadrunner-plugins/blob/master/http/middleware/new_relic/writer.go
Wrapper usage sample (make sure you use a `sync.Pool`): https://github.com/spiral/roadrunner-plugins/blob/master/http/middleware/new_relic/plugin.go#L77

If you need help, please, file a detailed feature request and I'll implement this middleware in a generic way (for all users).
